update README (#261)
phenyl-js/phenyl: README.md

Within these two environments, States are Synchronized in various ways.

# State Synchronization over Environments (SSoE)
State Synchronization over Environments is the main concept of modern applications.

> If you use RESTful API, you syncronize server and client state through HTTP methods.

Reviewer: ```suggestion
For example, if you use RESTful API, you can synchronize server and client state through following HTTP methods:
```

> Followings are examples.

Reviewer: ```suggestion

```

- Acquire server state of user by GET `/user` and provide user profile view by acquired user entity.
- User updates user name at profile view, app update user entity in the client side and app POST or PATCH `/user`.
- User delete account, app clear the client state and app execute DELETE `/user`

It's popular method of SSoE. So, many people create RDB, write API server to transform Data to Entity and communicate with client, write fetch code in client, client transform Entity to data for view, client update entity and notify to server, server save updated entity to DB and so on. Additionally error handlings are neccessary in practice.

> Phenyl provides more simple solution to handle SSoE base on 4 concepts.

Reviewer: ```suggestion
Phenyl provides simpler solution to handle SSoE based on the following 4 concepts:

 - State as one big JSON
 - OAD: Operations as Data
 - MongoDB-like operations
 - Git-like synchronization
```

> ## State is one big JSON

Reviewer: ```suggestion
## State as one big JSON
```

JSON is very useful format for JavaScript. It has some basic types like string or boolean and it's easy to serialize and deserialize.
We represent the entire app state as one big JSON.

## OAD: Operations As Data

> Phenyl handle all operations that like create an entity , update property in an entity or delete an entity as data. 

Reviewer: ```suggestion
Phenyl handles all CRUD operations on data, such as creating an entity, reading the data in an entity, updating a property in an entity, and deleting an entity as data. 
```

> All operations history is also the state of application.

Reviewer: ```suggestion
The history of all operations is also saved as in the state.
```

> This concept allow us easy to reproduce and debug any situations we want.

Reviewer: ```suggestion
This concept allows us to easily reproduce and debug any situations we want.
```

## MongoDB-like operations
Phenyl provides MongoDB-like operations.

> In mongoDB shell, example CRUD operations are like following.

Reviewer: ```suggestion
In a mongoDB shell, CRUD operations are performed as following:
```

```shell
> db.testUser.insertOne({_id: "test1", name: "Test1"})
{ "acknowledged" : true, "insertedId" : "test1" }

> db.testUser.findOne({_id: "test1"})
{ "_id" : "test1", "name" : "Test1" }

> db.testUser.updateOne({name: "Test1"}, {$set: {favoriteFood: "banana"}})
{ "acknowledged" : true, "matchedCount" : 1, "modifiedCount" : 1 } 

> db.testUser.findOne({_id: "test1"})
{ "_id" : "test1", "name" : "Test1" , "favoriteFood": "banana" }

> db.testUser.deleteOne({"_id": "test1"})
{ "acknowledged" : true, "deletedCount" : 1 } 

> db.testUser.findOne({_id: "test1"})
null
```

> In Phenyl client, the operations are like following. (This code is abbreviated.)

Reviewer: ```suggestion
In Phenyl client, the operations are performed as following. (This code is simplified.)
```

```ts
// omission prepare httpClient and preSession
const inserted = await phenylClient.insertAndGet(
    {
        entityName: "testUser",
        value: {
            id: "test1",
            name: "Test1",
        }
    },
    preSession.id
);

console.log(inserted.entity) 
// -> { id: "test1", name: "Test1" }

const updated = await phenylClient.updateAndGet(
    {
        entityName: "testUser",
        id: "test1",
        operation: { 
            $set: { favoriteFood: "banana" } 
        }
    },
    preSession.id
);

console.log(updated.entity) 
// -> { id: "test1", name: "Test1" , faboriteFood: "banana" }

await phenylClient.delete({ entityName: "testUser", id: "test1" })
```

If you have used mongodb, you will soon get to be friendly with the API of Phenyl😎.

## Git-like synchronization

> Phenyl synchronize between server and client by git-like command.

Reviewer: ```suggestion
Phenyl synchronizes between server and client by using git-like command.
```

Phenyl is powered by [sp2](https://github.com/phenyl-js/sp2), a set of JavaScript modules used for state operations. 

# Usage

> Phenyl needs you to implement 2 features, one is GeneralTypeMap and the other is functionalGroup.

Reviewer: ```suggestion
Phenyl needs you to implement 2 features, one is **GeneralTypeMap** and the other is **functionalGroup**.
```

> GeneralTypeMap is type definition that describes request and response of each entity and auth information.

Reviewer: ```suggestion
**GeneralTypeMap** is type definition that describes shape of request and response of each entity and auth information.
```

> functionalGroup is implementation to notify Phenyl about the domain that we want to use. 

Reviewer: ```suggestion
**functionalGroup** is implementation to notify Phenyl about the domain that we want to use. 
```
